File: app/src/main/java/pro/anton/averin/networking/testrest/presenters/RxBusPresenter.java
package pro.anton.averin.networking.testrest.presenters;

import javax.inject.Inject;

import pro.anton.averin.networking.testrest.BaseContext;
import pro.anton.averin.networking.testrest.rx.HasRxBusEvent;
import pro.anton.averin.networking.testrest.rx.RxBus;
import pro.anton.averin.networking.testrest.rx.events.RxBusEvent;
import pro.anton.averin.networking.testrest.utils.LLogger;

public class RxBusPresenter<B extends BaseView> extends BasePresenterImpl<B> implements HasRxBusEvent {

    @Inject
    RxBus rxBus;
    @Inject
    LLogger llogger;

    public RxBusPresenter(BaseContext baseContext) {
        super(baseContext);
    }

    @Override
    public void onResume() {
        super.onResume();
        if (isVisible) {
            rxBus.subscribe(this);
        }
    }

    @Override
    public void onVisible() {
        super.onVisible();
        rxBus.subscribe(this);
    }

    @Override
    public void onHidden() {
        super.onHidden();
        rxBus.unsubscribe(this);
    }

    @Override
    public void onPause() {
        super.onPause();
        rxBus.unsubscribe(this);
    }

    @Override
    public void onEvent(RxBusEvent event) {

    }
}

File: app/src/main/java/pro/anton/averin/networking/testrest/presenters/RxBusPresenter.kt
package pro.anton.averin.networking.testrest.presenters

import javax.inject.Inject

import pro.anton.averin.networking.testrest.BaseContext
import pro.anton.averin.networking.testrest.rx.GlobalBus
import pro.anton.averin.networking.testrest.utils.LLogger
import rx.Subscription
import rx.functions.Action1

abstract class RxBusPresenter<B : BaseView>(baseContext: BaseContext) : BasePresenterImpl<B>(baseContext) {

    @Inject
    lateinit var globalBus: GlobalBus
    @Inject
    lateinit var llogger: LLogger

    private var subscription: Subscription? = null

    override fun onResume() {
        super.onResume()
        if (isVisible) {
            subscribe()
        }
    }

    override fun onVisible() {
        super.onVisible()
        subscribe()
    }

    override fun onHidden() {
        super.onHidden()
        unsubscribe()
    }

    override fun onPause() {
        super.onPause()
        unsubscribe()
    }

    private fun subscribe() {
        subscription = globalBus.events().subscribe {
            onEvent(it)
        }

    }

    protected abstract fun onEvent(event: Any)

    private fun unsubscribe() {
        subscription?.unsubscribe()
    }
}
